File: shared/ts/src/messages.ts
import { CUSTOM_CATEGORY, isPresetCategory, isValidSlug, type Category } from "./categories.js";
import { MAX_TRACK_SLOT, MessageType } from "./constants.js";
import { isErrorCode, ProtocolError, type ErrorCode } from "./errors.js";

const encoder = new TextEncoder();
const decoder = new TextDecoder("utf-8", { fatal: true });

// ---------- HELLO ----------

export interface HelloPayload {
  version: 1;
  client: string;
  audio: {
    codec: "opus";
    sampleRate: 48000;
    channels: 1 | 2;
    frameMs: 20;
  };
}

export function encodeHello(p: HelloPayload): Uint8Array {
  return encodeJson({
    version: p.version,
    client: p.client,
    audio: {
      codec: p.audio.codec,
      sample_rate: p.audio.sampleRate,
      channels: p.audio.channels,
      frame_ms: p.audio.frameMs,
    },
  });
}

export function decodeHello(buf: Uint8Array): HelloPayload {
  const obj = decodeJson(buf);
  if (obj.version !== 1) throw bad("hello.version");
  const client = strField(obj, "client", 128);
  const audio = obj.audio;
  if (!audio || typeof audio !== "object") throw bad("hello.audio");
  if (audio.codec !== "opus") throw bad("hello.audio.codec");
  if (audio.sample_rate !== 48000) throw bad("hello.audio.sample_rate");
  if (audio.channels !== 1 && audio.channels !== 2) throw bad("hello.audio.channels");
  if (audio.frame_ms !== 20) throw bad("hello.audio.frame_ms");
  return {
    version: 1,
    client,
    audio: { codec: "opus", sampleRate: 48000, channels: audio.channels, frameMs: 20 },
  };
}

// ---------- TRACK_LIST ----------

export interface TrackInfo {
  id: number;
  slug: string;
  category: Category;
  label: string;
}

export interface TrackListPayload {
  tracks: TrackInfo[];
}

export function encodeTrackList(p: TrackListPayload): Uint8Array {
  return encodeJson({ tracks: p.tracks });
}

export function decodeTrackList(buf: Uint8Array): TrackListPayload {
  const obj = decodeJson(buf);
  if (!Array.isArray(obj.tracks)) throw bad("track_list.tracks");
  if (obj.tracks.length > MAX_TRACK_SLOT) throw bad("track_list too many tracks");

  const tracks: TrackInfo[] = [];
  const seen = new Set<number>();
  for (const raw of obj.tracks) {
    if (!raw || typeof raw !== "object") throw bad("track entry");
    const id = raw.id;
    if (!Number.isInteger(id) || id < 1 || id > MAX_TRACK_SLOT) throw bad("track.id");
    if (seen.has(id)) throw bad("track.id duplicate");
    seen.add(id);

    const slug = strField(raw, "slug", 64);
    if (!isValidSlug(slug)) throw bad("track.slug format");

    const category = strField(raw, "category", 32);
    if (category !== CUSTOM_CATEGORY && !isPresetCategory(category)) throw bad("track.category");

    const label = strField(raw, "label", 64);
    tracks.push({ id, slug, category: category as Category, label });
  }
  return { tracks };
}

// ---------- FINGERPRINT ----------

export interface FingerprintPayload {
  hash: bigint;
  windowMs: number;
}

export function encodeFingerprint(p: FingerprintPayload): Uint8Array {
  const buf = new Uint8Array(10);
  const view = new DataView(buf.buffer);
  view.setBigUint64(0, p.hash, false);
  view.setUint16(8, p.windowMs, false);
  return buf;
}

export function decodeFingerprint(buf: Uint8Array): FingerprintPayload {
  if (buf.length !== 10) throw bad("fingerprint length");
  const view = new DataView(buf.buffer, buf.byteOffset, buf.byteLength);
  return { hash: view.getBigUint64(0, false), windowMs: view.getUint16(8, false) };
}

// ---------- TRACK_META ----------

export interface TrackMetaPayload {
  title?: string;
  artist?: string;
  albumArtUrl?: string;
}

export function encodeTrackMeta(p: TrackMetaPayload): Uint8Array {
  const out: Record<string, string> = {};
  if (p.title !== undefined) out.title = p.title;
  if (p.artist !== undefined) out.artist = p.artist;
  if (p.albumArtUrl !== undefined) out.album_art_url = p.albumArtUrl;
  return encodeJson(out);
}

export function decodeTrackMeta(buf: Uint8Array): TrackMetaPayload {
  const obj = decodeJson(buf);
  const out: TrackMetaPayload = {};
  if (obj.title !== undefined) out.title = strField(obj, "title", 256);
  if (obj.artist !== undefined) out.artist = strField(obj, "artist", 256);
  if (obj.album_art_url !== undefined) {
    const u = strField(obj, "album_art_url", 1024);
    if (!u.startsWith("https://")) throw bad("album_art_url must be https");
    out.albumArtUrl = u;
  }
  return out;
}

// ---------- STATS ----------

export interface StatsPayload {
  uptimeS: number;
  packetsSent: number;
  subscribers: number;
  tracksActive: number;
}

export function encodeStats(p: StatsPayload): Uint8Array {
  return encodeJson({
    uptime_s: p.uptimeS,
    packets_sent: p.packetsSent,
    subscribers: p.subscribers,
    tracks_active: p.tracksActive,
  });
}

export function decodeStats(buf: Uint8Array): StatsPayload {
  const obj = decodeJson(buf);
  return {
    uptimeS: numField(obj, "uptime_s"),
    packetsSent: numField(obj, "packets_sent"),
    subscribers: numField(obj, "subscribers"),
    tracksActive: numField(obj, "tracks_active"),
  };
}

// ---------- SUB_TRACKS ----------

export interface SubTracksPayload {
  tracks: number[];
}

export function encodeSubTracks(p: SubTracksPayload): Uint8Array {
  return encodeJson({ tracks: p.tracks });
}

export function decodeSubTracks(buf: Uint8Array): SubTracksPayload {
  const obj = decodeJson(buf);
  if (!Array.isArray(obj.tracks)) throw bad("sub_tracks");
  if (obj.tracks.length > MAX_TRACK_SLOT) throw bad("sub_tracks too many");
  const tracks: number[] = [];
  const seen = new Set<number>();
  for (const id of obj.tracks) {
    if (!Number.isInteger(id) || id < 1 || id > MAX_TRACK_SLOT) throw bad("sub_tracks.id");
    if (seen.has(id)) throw bad("sub_tracks.id duplicate");
    seen.add(id);
    tracks.push(id);
  }
  return { tracks };
}

// ---------- ERROR ----------

export interface ErrorPayload {
  code: ErrorCode;
  message: string;
}

export function encodeError(p: ErrorPayload): Uint8Array {
  return encodeJson({ code: p.code, message: p.message });
}

export function decodeError(buf: Uint8Array): ErrorPayload {
  const obj = decodeJson(buf);
  const code = strField(obj, "code", 32);
  if (!isErrorCode(code)) throw bad("error.code");
  const message = strField(obj, "message", 512);
  return { code, message };
}

// ---------- helpers ----------

const MAX_JSON_BYTES = 4096;

function encodeJson(value: unknown): Uint8Array {
  const out = encoder.encode(JSON.stringify(value));
  if (out.length > MAX_JSON_BYTES) {
    throw new ProtocolError(`json payload too large: ${out.length}`, "TOO_LARGE");
  }
  return out;
}

function decodeJson(buf: Uint8Array): Record<string, unknown> & { [k: string]: any } {
  if (buf.length > MAX_JSON_BYTES) {
    throw new ProtocolError(`json payload too large: ${buf.length}`, "TOO_LARGE");
  }
  let parsed: unknown;
  try {
    parsed = JSON.parse(decoder.decode(buf));
  } catch {
    throw new ProtocolError("invalid json or utf-8", "BAD_PAYLOAD");
  }
  if (!parsed || typeof parsed !== "object" || Array.isArray(parsed)) {
    throw new ProtocolError("payload must be a json object", "BAD_PAYLOAD");
  }
  return parsed as Record<string, unknown> & { [k: string]: any };
}

function strField(obj: Record<string, unknown>, key: string, maxLen: number): string {
  const v = obj[key];
  if (typeof v !== "string") throw bad(`field ${key} must be string`);
  if (v.length > maxLen) throw bad(`field ${key} too long`);
  return v;
}

function numField(obj: Record<string, unknown>, key: string): number {
  const v = obj[key];
  if (typeof v !== "number" || !Number.isFinite(v) || v < 0) throw bad(`field ${key} must be non-negative finite number`);
  return v;
}

function bad(detail: string): ProtocolError {
  return new ProtocolError(`malformed payload: ${detail}`, "BAD_PAYLOAD");
}

// Re-export MessageType for callers that want to dispatch.
export { MessageType };

File: shared/go/messages.go
package wire

import (
	"encoding/binary"
	"encoding/json"
	"strings"
	"unicode/utf8"
)

const maxJSONBytes = 4096

// ---------- HELLO ----------

type Hello struct {
	Version int    `json:"version"`
	Client  string `json:"client"`
	Audio   struct {
		Codec      string `json:"codec"`
		SampleRate int    `json:"sample_rate"`
		Channels   int    `json:"channels"`
		FrameMS    int    `json:"frame_ms"`
	} `json:"audio"`
}

func EncodeHello(h Hello) ([]byte, error) { return encJSON(h) }

func DecodeHello(buf []byte) (Hello, error) {
	var h Hello
	if err := decJSON(buf, &h); err != nil {
		return Hello{}, err
	}
	if h.Version != 1 {
		return Hello{}, bad("hello.version")
	}
	if err := checkStr("client", h.Client, 128); err != nil {
		return Hello{}, err
	}
	if h.Audio.Codec != "opus" {
		return Hello{}, bad("hello.audio.codec")
	}
	if h.Audio.SampleRate != 48000 {
		return Hello{}, bad("hello.audio.sample_rate")
	}
	if h.Audio.Channels != 1 && h.Audio.Channels != 2 {
		return Hello{}, bad("hello.audio.channels")
	}
	if h.Audio.FrameMS != 20 {
		return Hello{}, bad("hello.audio.frame_ms")
	}
	return h, nil
}

// ---------- TRACK_LIST ----------

type TrackInfo struct {
	ID       uint8  `json:"id"`
	Slug     string `json:"slug"`
	Category string `json:"category"`
	Label    string `json:"label"`
}

type TrackList struct {
	Tracks []TrackInfo `json:"tracks"`
}

func EncodeTrackList(t TrackList) ([]byte, error) { return encJSON(t) }

func DecodeTrackList(buf []byte) (TrackList, error) {
	var t TrackList
	if err := decJSON(buf, &t); err != nil {
		return TrackList{}, err
	}
	if len(t.Tracks) > MaxTrackSlot {
		return TrackList{}, bad("track_list too many tracks")
	}
	seen := make(map[uint8]bool, len(t.Tracks))
	for _, tr := range t.Tracks {
		if tr.ID < 1 || tr.ID > MaxTrackSlot {
			return TrackList{}, bad("track.id")
		}
		if seen[tr.ID] {
			return TrackList{}, bad("track.id duplicate")
		}
		seen[tr.ID] = true
		if err := checkStr("track.slug", tr.Slug, 64); err != nil {
			return TrackList{}, err
		}
		if !IsValidSlug(tr.Slug) {
			return TrackList{}, bad("track.slug format")
		}
		if tr.Category != CustomCategory && !IsPresetCategory(tr.Category) {
			return TrackList{}, bad("track.category")
		}
		if err := checkStr("track.label", tr.Label, 64); err != nil {
			return TrackList{}, err
		}
	}
	return t, nil
}

// ---------- FINGERPRINT ----------

type Fingerprint struct {
	Hash     uint64
	WindowMS uint16
}

func EncodeFingerprint(f Fingerprint) []byte {
	buf := make([]byte, 10)
	binary.BigEndian.PutUint64(buf[0:8], f.Hash)
	binary.BigEndian.PutUint16(buf[8:10], f.WindowMS)
	return buf
}

func DecodeFingerprint(buf []byte) (Fingerprint, error) {
	if len(buf) != 10 {
		return Fingerprint{}, bad("fingerprint length")
	}
	return Fingerprint{
		Hash:     binary.BigEndian.Uint64(buf[0:8]),
		WindowMS: binary.BigEndian.Uint16(buf[8:10]),
	}, nil
}

// ---------- TRACK_META ----------

type TrackMeta struct {
	Title       string `json:"title,omitempty"`
	Artist      string `json:"artist,omitempty"`
	AlbumArtURL string `json:"album_art_url,omitempty"`
}

func EncodeTrackMeta(m TrackMeta) ([]byte, error) { return encJSON(m) }

func DecodeTrackMeta(buf []byte) (TrackMeta, error) {
	var m TrackMeta
	if err := decJSON(buf, &m); err != nil {
		return TrackMeta{}, err
	}
	if err := checkStr("title", m.Title, 256); err != nil {
		return TrackMeta{}, err
	}
	if err := checkStr("artist", m.Artist, 256); err != nil {
		return TrackMeta{}, err
	}
	if err := checkStr("album_art_url", m.AlbumArtURL, 1024); err != nil {
		return TrackMeta{}, err
	}
	if m.AlbumArtURL != "" && !strings.HasPrefix(m.AlbumArtURL, "https://") {
		return TrackMeta{}, bad("album_art_url must be https")
	}
	return m, nil
}

// ---------- STATS ----------

type Stats struct {
	UptimeS      int64 `json:"uptime_s"`
	PacketsSent  int64 `json:"packets_sent"`
	Subscribers  int64 `json:"subscribers"`
	TracksActive int64 `json:"tracks_active"`
}

func EncodeStats(s Stats) ([]byte, error) { return encJSON(s) }

func DecodeStats(buf []byte) (Stats, error) {
	var s Stats
	if err := decJSON(buf, &s); err != nil {
		return Stats{}, err
	}
	if s.UptimeS < 0 || s.PacketsSent < 0 || s.Subscribers < 0 || s.TracksActive < 0 {
		return Stats{}, bad("stats: non-negative required")
	}
	return s, nil
}

// ---------- SUB_TRACKS ----------

type SubTracks struct {
	Tracks []uint8 `json:"tracks"`
}

func EncodeSubTracks(s SubTracks) ([]byte, error) { return encJSON(s) }

func DecodeSubTracks(buf []byte) (SubTracks, error) {
	var s SubTracks
	if err := decJSON(buf, &s); err != nil {
		return SubTracks{}, err
	}
	if len(s.Tracks) > MaxTrackSlot {
		return SubTracks{}, bad("sub_tracks too many")
	}
	seen := make(map[uint8]bool, len(s.Tracks))
	for _, id := range s.Tracks {
		if id < 1 || id > MaxTrackSlot {
			return SubTracks{}, bad("sub_tracks.id")
		}
		if seen[id] {
			return SubTracks{}, bad("sub_tracks.id duplicate")
		}
		seen[id] = true
	}
	return s, nil
}

// ---------- ERROR ----------

type ErrorPayload struct {
	Code    string `json:"code"`
	Message string `json:"message"`
}

func EncodeError(e ErrorPayload) ([]byte, error) { return encJSON(e) }

func DecodeError(buf []byte) (ErrorPayload, error) {
	var e ErrorPayload
	if err := decJSON(buf, &e); err != nil {
		return ErrorPayload{}, err
	}
	if !IsErrorCode(e.Code) {
		return ErrorPayload{}, bad("error.code")
	}
	if err := checkStr("error.message", e.Message, 512); err != nil {
		return ErrorPayload{}, err
	}
	return e, nil
}

// ---------- helpers ----------

func encJSON(v any) ([]byte, error) {
	out, err := json.Marshal(v)
	if err != nil {
		return nil, protoErr(BadPayload, "json marshal: %v", err)
	}
	if len(out) > maxJSONBytes {
		return nil, protoErr(TooLarge, "json payload too large: %d", len(out))
	}
	return out, nil
}

func decJSON(buf []byte, v any) error {
	if len(buf) > maxJSONBytes {
		return protoErr(TooLarge, "json payload too large: %d", len(buf))
	}
	if !utf8.Valid(buf) {
		return protoErr(BadPayload, "invalid utf-8")
	}
	dec := json.NewDecoder(strings.NewReader(string(buf)))
	dec.DisallowUnknownFields()
	if err := dec.Decode(v); err != nil {
		return protoErr(BadPayload, "json: %v", err)
	}
	// reject trailing junk
	if dec.More() {
		return protoErr(BadPayload, "trailing json data")
	}
	return nil
}

func checkStr(field, v string, maxLen int) error {
	if len(v) > maxLen {
		return protoErr(BadPayload, "field %s too long", field)
	}
	return nil
}

func bad(detail string) error {
	return protoErr(BadPayload, "%s", detail)
}

